homeassistant/components/yeelight/__init__.py

```diff
@@ -139,73 +148,192 @@
-        platform_config = hass.data[DATA_YEELIGHT][ipaddr].config.copy()
-        platform_config[CONF_HOST] = ipaddr
-        platform_config[CONF_CUSTOM_EFFECTS] = config.get(DOMAIN, {}).get(
-            CONF_CUSTOM_EFFECTS, {}
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
         )
-        load_platform(hass, LIGHT_DOMAIN, DOMAIN, platform_config, config)
-        load_platform(hass, BINARY_SENSOR_DOMAIN, DOMAIN, platform_config, config)
+    )
 
-    dispatcher_connect(hass, DEVICE_INITIALIZED, load_platforms)
+    if unload_ok:
+        device = hass.data[DOMAIN][DATA_CONFIG_ENTRIES].pop(entry.entry_id)
+        device.async_unload()
+        if entry.data[CONF_ID]:
+            # discovery
+            scanner = await YeelightScanner.async_get(hass)
+            await scanner.async_unregister_callback(entry.data[CONF_ID])
 
-    if DOMAIN in config:
-        for ipaddr, device_config in conf[CONF_DEVICES].items():
-            _LOGGER.debug("Adding configured %s", device_config[CONF_NAME])
-            _setup_device(hass, config, ipaddr, device_config)
+    return unload_ok
 
-    return True
+
+async def _async_setup_device(hass: HomeAssistant, ipaddr: str, config: dict,) -> None:
+    # Set up device
+    bulb = Bulb(ipaddr, model=config.get(CONF_MODEL) or None)
+    capabilities = await hass.async_add_executor_job(bulb.get_capabilities)
+    if capabilities is None:  # timeout
+        _LOGGER.error("Failed to get capabilities from %s", ipaddr)
+        raise ConfigEntryNotReady
+    device = YeelightDevice(hass, ipaddr, config, bulb)
+    await hass.async_add_executor_job(device.update)
+    await device.async_setup()
+    return device
 
 
-def _setup_device(hass, _, ipaddr, device_config):
-    devices = hass.data[DATA_YEELIGHT]
+async def _async_update_listener(hass: HomeAssistant, entry: ConfigEntry):
+    """Handle options update."""
+    await hass.config_entries.async_reload(entry.entry_id)
 
-    if ipaddr in devices:
-        return
 
-    device = YeelightDevice(hass, ipaddr, device_config)
+class YeelightScanner:
+    """Scan for Yeelight devices."""
 
-    devices[ipaddr] = device
-    hass.add_job(device.setup)
+    _scanner = None
+    _get_lock = asyncio.Lock()
+
+    @staticmethod
+    async def async_get(hass: HomeAssistant):
+        """Get scanner instance."""
+        with await YeelightScanner._get_lock:
+            if YeelightScanner._scanner is None:
+                YeelightScanner._scanner = YeelightScanner(hass)
+        return YeelightScanner._scanner
+
+    def __init__(self, hass: HomeAssistant):
+        """Initialize class."""
+        self._hass = hass
+        self._seen = {}
+        self._seen_lock = asyncio.Lock()
```

Comment:
We don't need this lock either since we don't yield to the event loop inside the lock.
```suggestion
```